### PR TITLE
[WRKLDS-730] refactor route controller manager to use library-go server and ControlerCommand for generic setup

### PR DESCRIPTION
**Pull request by atiratree**
and simplify ControllerContext.

CI is expected to fail until we get the operator PR in (https://github.com/openshift/route-controller-manager/pull/22#issuecomment-1540423857)

TODO:

- [x] https://github.com/openshift/cluster-openshift-controller-manager-operator/pull/288
- [x] https://github.com/openshift/route-controller-manager/pull/21

### CODE REVIEW

**Comment by atiratree** [comment]
The CI passes even without operator PR being merged. In RCM logs we can that the leader election obtains an incorrect default lock - this could cause a race if an old instance of RCM would use the original `openshift-route-controllers` lock. We still have to wait for the operator PR as this race is not detected by CI. After that we should check the logs/events before merging this PR to ensure the correct lock is being used.

```
event.go:298] Event(v1.ObjectReference{Kind:"Lease", Namespace:"openshift-route-controller-manager", Name:"route-controller-manager", Name:"route-controller-manager-lock", UID:"9946ed27-bfae-4184-a7e3-8949241bc8f2", APIVersion:"coordination.k8s.io/v1", ResourceVersion:"23162", FieldPath:""}): type: 'Normal' reason: 'LeaderElection' route-controller-78cd8495c7-kxr9d_5496b873-aff5-4ffa-a591-15f345bb12c7 became leader 
```

**Comment by atiratree** [comment]
/test all
to consume changes from https://github.com/openshift/cluster-openshift-controller-manager-operator/pull/288, then manually confirm (https://github.com/openshift/route-controller-manager/pull/22#issuecomment-1540423857)

**Comment by atiratree** [comment]
correct lease is being picked up in the recent [run](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_route-controller-manager/22/pull-ci-openshift-route-controller-manager-main-e2e-aws-ovn/1660671022115852288)
```
event.go:298] Event(v1.ObjectReference{Kind:"Lease", Namespace:"openshift-route-controller-manager", Name:"openshift-route-controllers", UID:"32ce9fd8-9a8b-4b65-8531-db4b48f908ec", APIVersion:"coordination.k8s.io/v1", ResourceVersion:"20139", FieldPath:""}): type: 'Normal' reason: 'LeaderElection' route-controller-manager-867999cd9b-xjrxk_baa8fad5-3dbb-4b65-af4b-d0d848a092be became leader
```

File: pkg/cmd/controller/route/interfaces.go [hunk 1]
```diff
@@ -5,58 +5,32 @@ import (
...
-	OpenshiftConfigClient(name string) (configclient.Interface, error)
-	OpenshiftConfigClientOrDie(name string) configclient.Interface
-
-	OpenshiftOperatorClient(name string) (operatorclient.Interface, error)
-	OpenshiftOperatorClientOrDie(name string) operatorclient.Interface
 }
 
-type ControllerContext struct {
-	// TODO: Make this minimal config instead of passing entire controller manager config
+type EnhancedControllerContext struct {
```
Comment (ingvagabund) [inline pkg/cmd/controller/route/interfaces.go]: Why the rename from `ControllerContext` to `EnhancedControllerContext`? Is `ControllerContext` no longer a valid keyword?

File: pkg/cmd/controller/route/standalone_apiserver.go [hunk 2]
```diff
@@ -1,158 +0,0 @@
...
-	authzwebhook "k8s.io/apiserver/plugin/pkg/authorizer/webhook"
-	clientgoclientset "k8s.io/client-go/kubernetes"
-	"k8s.io/kubernetes/pkg/api/legacyscheme"
-
-	configv1 "github.com/openshift/api/config/v1"
-	"github.com/openshift/library-go/pkg/crypto"
-)
-
-// TODO make this an actual API server built on the genericapiserver
-func RunControllerServer(servingInfo configv1.HTTPServingInfo, kubeExternal clientgoclientset.Interface) error {
```
Comment (ingvagabund) [inline pkg/cmd/controller/route/standalone_apiserver.go:30]: `RunControllerServer` is no longer needed? Has it been ever used before?

File: pkg/cmd/route-controller-manager/controller_manager.go [hunk 3]
```diff
@@ -2,32 +2,50 @@ package openshift_controller_manager
...
-	serviceability.InitLogrusFromKlog()
-	kubeClient, err := kubernetes.NewForConfig(clientConfig)
+func RunRouteControllerManager(ctx context.Context, controllerContext *controllercmd.ControllerContext) error {
+	config, err := asOpenshiftControllerManagerConfig(controllerContext.ComponentConfig)
 	if err != nil {
 		return err
 	}
 
-	// only serve if we have serving information.
-	if config.ServingInfo != nil {
```
Comment (ingvagabund) [inline pkg/cmd/route-controller-manager/controller_manager.go:25]: Has `ServingInfo` ever been set?

File: pkg/cmd/controller/route/interfaces.go (hunk 1 above)
Comment (atiratree) [reply pkg/cmd/controller/route/interfaces.go]: to distinguish it from the embedded `controllercmd.ControllerContext`  - we will need the embedded one in later refactoring when accessing for example the `EventRecorder`

File: pkg/cmd/route-controller-manager/controller_manager.go (hunk 3 above)
Comment (atiratree) [reply pkg/cmd/route-controller-manager/controller_manager.go:25]: probably not from the outside, but we are calling `SetRecommendedHTTPServingInfoDefaults` which initializes it

File: pkg/cmd/route-controller-manager/cmd.go [hunk 4]
```diff
@@ -2,119 +2,38 @@ package openshift_controller_manager
...
-	// Resolve relative to CWD
-	absoluteConfigFile, err := api.MakeAbs(o.ConfigFilePath, "")
-	if err != nil {
-		return err
-	}
-	configFileLocation := path.Dir(absoluteConfigFile)
-
-	config := obj.(*openshiftcontrolplanev1.OpenShiftControllerManagerConfig)
-	/// this isn't allowed to be nil when by itself.
-	// TODO remove this when the old path is gone.
```
[the remark this answers is not in the document]
Comment (atiratree) [reply pkg/cmd/route-controller-manager/cmd.go:106]: I think this is a legacy TODO, when the code was still in origin

File: pkg/cmd/controller/route/standalone_apiserver.go (hunk 2 above)
Comment (atiratree) [reply pkg/cmd/controller/route/standalone_apiserver.go:30]: it was run before, now it is responsibility of the library-go

**Pull request merged**